src/Microsoft.ML.Transforms/StatefulCustomMappingTransformer.cs

```diff
@@ -0,0 +1,398 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.ML.Data;
+using Microsoft.ML.Internal.Utilities;
+using Microsoft.ML.Runtime;
+
+namespace Microsoft.ML.Transforms
+{
+    /// <summary>
+    /// <see cref="ITransformer"/> resulting from fitting an <see cref="StatefulCustomMappingEstimator{TSrc, TState, TDst}"/>.
+    /// </summary>
+    /// <typeparam name="TSrc">The class defining which columns to take from the incoming data.</typeparam>
+    /// <typeparam name="TDst">The class defining which new columns are added to the data.</typeparam>
+    /// <typeparam name="TState">The type that describes per-cursor state.</typeparam>
+    public sealed class StatefulCustomMappingTransformer<TSrc, TDst, TState> : ITransformer
+        where TSrc : class, new()
+        where TDst : class, new()
+        where TState : class, new()
+    {
+        private readonly IHost _host;
+        private readonly Action<TSrc, TDst, TState> _mapAction;
+        private readonly Action<TState> _stateInitAction;
+        private readonly string _contractName;
+
+        internal InternalSchemaDefinition AddedSchema { get; }
+        internal SchemaDefinition InputSchemaDefinition { get; }
+
+        /// <summary>
+        /// Whether a call to <see cref="ITransformer.GetRowToRowMapper(DataViewSchema)"/> should succeed, on an
+        /// appropriate schema.
+        /// </summary>
+        bool ITransformer.IsRowToRowMapper => true;
```

Comment:
I have one question: If this is `RowToRowMapper`, then, according to what we chatted the other day, this can actually be used with `PredictionEngine`, right? But my understanding was that `PredictionEngine` process 1 row at a time. So I wonder: what will happen to the "state" on this transform? Will it actually update the state accross rows or will it restart the state on every row with the init function?

Reply:
The init function gets called once when the getter is created, and since it is created only once when `PredictionEngine` is created, then it will use the same state object, that changes from prediction to prediction.

---
In reply to: [401998841](https://github.com/dotnet/machinelearning/pull/4676#discussion_r401998841) [](ancestors = 401998841)